### ISSUE TITLE
Is Avahi abandoned? New maintainer?

### ISSUE THREAD

**Issue by adriancable** [lathiat/avahi#388]
@lathiat - can you confirm that you have now abandoned further development and maintenance of Avahi, as appears to be the case?

While I interpret no response from you = confirmation of abandonment, judging by the number of recent PRs submitted I would urge you to say so definitively. Otherwise people are putting in a significant amount of work to PRs that will never go anywhere. This is disrespectful to others' time, especially since a 1-line post here saying 'I have abandoned Avahi' would take less than 20 seconds of yours.

If you have abandoned Avahi, please tell us. In that case, also please tell us whether you would be willing to hand the repo over to a new maintainer, should someone volunteer (which I think is highly likely).

**Comment by lathiat** [comment]
It’s not abandoned but there is a solid shortage of time spent currently that I hope to rectify. 

I’m open to contributing maintainers joining on

**Comment by adriancable** [comment]
@lathiat - that's great to know - thank you.

If it's helpful, I volunteer to be a maintainer. I do not have much experience with the avahi-autoipd side, however, I am the maintainer of a private fork of avahi-daemon which is used on a commercial IoT device. This fork incorporates a number of open PRs on this repo, along with other fixes (principally to work around IPv6 router bugs), which address commonplace issues and is running on a large number of devices 'in the field'. So I can at least vouch for specific PRs as being effective and unlikely to introduce side effects, and/or submit new PRs with some additional fixes, if this is helpful.

**Comment by lathiat** [comment]
That sounds great. Can you shoot me an e-mail with a way to get in touch with you (twitter/irc/something) so we can have a chat?

Reviewing/testing the PRs is one of the bottle necks so if nothing else if theres something you've validated in the field im happy to take that as a +1 to just merge it.

**Comment by adriancable** [comment]
Sure thing! Happy to go through what I’m running in the field with you, and so we can pick some low hanging fruit / well tested PRs to get merged.

**Comment by EskoDijk** [comment]
@adriancable Maybe one quick action to take now is to label #243 not as an enhancement, but as a bug!  This issue keeps hitting people in my development projects. And the specifications are quite clear I think on what the implementation should do.  And for the same reasoning #269 can be rejected, I believe.

**Comment by pemensik** [comment]
Hi, I would like to contribute if it was possible. But I think this is not the best communication platform. Is there a reason, why is [avahi mailing list](https://lists.freedesktop.org/mailman/listinfo/avahi) dead? I would volunteer to occasional moderation if needed. But can I suggest opening that mailing list as a start? Some place we could discuss changes would be nice. Can you please open mailing list to automatic registration, because @lathiat clearly does not have time to manage it manually.

**Comment by pemensik** [comment]
Thanks for opening the mailing list. It is now open to new people, tested: https://lists.freedesktop.org/archives/avahi/2022-October/thread.html

**Comment by Neustradamus** [comment]
It is possible to move in an organization place to the future developement?
- https://github.com/avahi-project

I have already invited @lathiat to move the repository.

Who can join it?

Note, I have done a long time ago:
- https://github.com/lathiat/avahi/issues/245
- https://github.com/lathiat/avahi/issues/258
- https://github.com/lathiat/avahi/issues/325
- https://github.com/lathiat/avahi/issues/335

**Comment by Neustradamus** [comment]
@lathiat: Can you do the transfer for the future?

I think that you can do for nss-mdns too.

cc: @adriancable and @pemensik.

**Comment by pemensik** [comment]
I have prepared a list of candidate bugs on wiki page. Most of them looks just right. Can somebody else look at them, maybe rate their importance too? I think we definitely want to merge the CVE soon. Is there anyone else, who can assign labels to PR and issues?

**Comment by Neustradamus** [comment]
@lathiat: After several weeks, several months, it is possible to look for the move/transfer?

It is really easy to do it, at the bottom:
- https://github.com/lathiat/avahi/settings
- https://github.com/lathiat/nss-mdns/settings

Push the "Transfer" button...

Thanks in advance.

**Comment by EskoDijk** [comment]
@pemensik You mention a list of bugs on the Wiki - do you mean the links to the list of patches there? (Or should I be looking elsewhere?)  I could label issues but the project would need to be located somewhere with an active maintainer so that people can be given rights to e.g. apply labels, close issues, and such.  Maybe discuss further on the mailing list?

**Comment by pemensik** [comment]
I have just received commit rights. @adriancable  has them as well, as I can see. I will try to review pending pull requests. If you had any changes needed, please create a pull request if it does have some already. Ping me if you have anything important.

**Comment by Neustradamus** [comment]
@pemensik: Now we wait the transfer to the organization...
@lathiat: Do not forget to do it!

**Comment by evverx** [comment]
> I will try to review pending pull requests.

@pemensik I wonder if it would make sense to set up some sort of CI first. As far as I can see there are quite a few PRs waiting to be merged but they don't seem to be even compile-tested anywhere. It seems avahi was built on Travis CI back in the day but now that it's gone the CI is no longer functional. I mostly care about various static analyzers and fuzzers (and I opened https://github.com/google/oss-fuzz/pull/9107 to be able to keep the fuzz targets closer eventually) and can bring CodeQL and Coverity if anyone is interested but I think some CI jobs simply compiling avahi and running its three unit tests on various architectures (using say Packit) should probably be up and running before PRs get merged to avoid breaking builds downstream at least. Just my 2 cents.

**Comment by EskoDijk** [comment]
@evverx I agree that a CI would be very useful here. For example just the standard CI that Github offers already has a lot of features. Here's an example of a pretty advanced cross-platform build definition as part of standard Github CI: https://github.com/openthread/openthread/blob/main/.github/workflows/build.yml

**Comment by pemensik** [comment]
Yes, it turns the avahi code is not as straight forward as it seemed. At least basic checks done on pull requests would be nice. If we could create at least basic functional tests, it would be even better. There were some PR I marked with CI label. But I would create separate issue for it. I think CI is important enough to have separate issue. This issue were about entirely different thing.

**Comment by pemensik** [comment]
Created issue #412 to track automatic checks on PR action.

**Comment by evverx** [comment]
> But I would create separate issue for i

Makes sense.

> This issue were about entirely different thing.

I commented here because to set up some things like Packit for example it's necessary to be the owner of the repository so in a way it boils to down to whether there are active maintainers with this kind of access. My understanding is that all the new maintainers have only commit access and that isn't enough to change any repository settings affecting CI among other things.

> Here's an example of a pretty advanced cross-platform build definition as part of standard Github CI

@EskoDijk Thanks! I suggested Packit because there almost all architectures apart from armhfp and probably ppc64le are native. On GitHub almost all GHActions like that boil down to spawning QEMU VMs or multi-arch containers (usually boiling down to qemu-user as well) and depending on the project it can lead to annoying issues like https://gitlab.com/qemu-project/qemu/-/issues/1248. It's just easier to build stuff natively. Anyway let's move this discussion to #412.

**Comment by Neustradamus** [comment]
Hello all,

I do not arrive to speak with @lathiat, same to you?

He is on Twitter too: https://twitter.com/lathiat.

The future of Avahi and nss-mdns is now...

**Comment by evverx** [comment]
@Neustradamus I think all the questions regarding the state of the repository were answered on the mailing list in https://lists.freedesktop.org/archives/avahi/2022-December/002587.html.

**Comment by lathiat** [comment]
Yeah, that mailling list post was the main update. I am excited by the activity from various people :)

3 main outstanding requests of relevance:

1. Commit access requests from @evverx and @Neustradamus 
2. Moving the avahi and nss-mdns repositories into a GitHub org

(the combination of those two will assist the comment from @evverx needing admin access for packit etc)

I will handle those soon.

**Comment by evverx** [comment]
@Neustradamus it's probably a weird question but I'm just curious. I wonder if you are planning to review/merge PRs/change repository settings? I' just trying to figure out why commit access is necessary. Assuming https://github.com/avahi-project was created by you it would also mean that if this repository was transferred there you would have full admin access with access to advisories and stuff like that.

**Comment by Neustradamus** [comment]
The main goal is to have a nice future for avahi and nss-mdns!

Several projects have been already saved from the death.

For information, Debian 12 freeze is very soon.

**Comment by Neustradamus** [comment]
@lathiat: The devel will be better if avahi and nss-mdns will be in the new place!
- https://github.com/avahi-project

Several months have been lost...

Too late for Debian 12.

**Comment by evverx** [comment]
> Several projects have been already saved from the death.

I'm not sure how moving stuff around can help with anything. The problem is that at this point nobody (apart from @pemensik) tries to review PRs/triage and fix bugs and so on. On top of that there is no adequate testing infrastructure and that makes it really hard to actually do anything without introducing even more bugs. (I brought some pretty basic testing stuff but it isn't enough to test non-trivial features like monotonic timers for example).

I think that this repository should just be converted into a public archive to make it clear that it isn't maintained any more. Downstream consumers that claim that they actually support `avahi` can always fork it and if their fork isn't dead everyone can switch to it.

**Comment by evverx** [comment]
> I brought some pretty basic testing stuff but it isn't enough to test non-trivial features like monotonic timers for example

Just to expand on that I hooked up the repository to Coveralls in https://github.com/lathiat/avahi/pull/461 and https://coveralls.io/builds/60369370 shows that approximately 10% is covered at this point. What I'm trying to say is that in terms of saving avahi it would be much more important to start covering it with tests to make it easier for reviewers and contributors to make sure that stuff works and regressions aren't introduced. PRs are welcome as usual.

Another way to help would be to go over https://scan.coverity.com/projects/avahi-daemon and help to fix some of those issues. I can give access to it if anyone is interested.

**Comment by lathiat** [comment]
I have added @evverx as a contributor. @Neustradamus seems just motivated to keep the project going but not really to be involved in PR reviewing/merging so will not add them but if that changes let me know and we can discuss further.

**Comment by evverx** [comment]
@lathiat thanks! I think it would be great if @mbiebl could be invited too to represent Debian (if it's something @mbiebl is interested in of course). The "collaborator"/"member" tag helps with reviews in the sense that those reviews can be blocking for example. @EskoDijk helps with various RFC corner cases. Then again I'm not sure if @EskoDijk is interested in that. It's just some stuff based on my observations.

**Comment by Neustradamus** [comment]
@lathiat: Thanks to have added @evverx.

An exemple of the current problem, you can see that the code is in your personal account (Avahi and nss-mdns).

You can see that I have permit 2 releases for PPP and active development: https://github.com/ppp-project/ppp
And one release and a not abandoned developement for radvd: https://github.com/radvd-project/radvd

**Comment by evverx** [comment]
@Neustradamus I'm not saying that avahi should never move. I agree that it would be better if it landed elsewhere where it could have several owners to make some stuff easier (and I think it has been said here several times here). What I'm saying is that there is no need to keep repeating that here and in https://github.com/lathiat/avahi/issues/430 and in other places. Other than that it probably looks like it should be possible to just press one button and call it a day but unfortunately it isn't that easy. For example OSS-Fuzz should be pointed to that new place, some GitHub actions should be changed, https://fossies.org/linux/test/avahi-master.tar.gz/codespell.svg should be pointed to the new default branch and so on. Would you be willing to volunteer to do all that?

> 2 releases

There are several people who can release avahi here but there is nothing to release yet regardless of where avahi lives.

> PPP and active development

It's active because there are people who actively participate in its development.

Anyway my point is that instead of making avahi look alive it would be nice to actually make it alive by participating in code reviews. Without any sort of automation testing stuff manually is important too. And looking at for example https://github.com/lathiat/avahi/pull/229#issuecomment-1567790900 I don't think that several years in the field suffice. If anyone is interested in https://github.com/lathiat/avahi/pull/181 they can try to roll it out locally and make sure that the syscall filters don't break some unusual configurations (as they usually do) and so on.

**Comment by evverx** [comment]
> There are redirections

There are not in those particular cases. I'm not talking about `git clone` or browsing the repository on GitHub. I'm talking about stuff like `if repostiory == "lathiat/avahi"`.

**Comment by lathiat** [comment]
> @lathiat: There is a big problem, who has done a change in the GitHub Avahi that I have created for the project? There are not serious people? It is really strange! Now, I need to create a new organization for the future?

For the moment, there is nothing major being blocked by it existing in my personal namespace.

The primary issue is contributors to review and merge PRs. Both @pemensik and @evverx have both been graciously doing some great work on this which I (and I am sure everyone else) am very appreciative of. They have both now been given access to the current repository so they can do everything needed directly. There is also a great community of people that have been contributing patches of many different varieties.

You specifically have also been great at keeping me accountable and pushing for more action to take place (which has been happening) and I am also thankful for that. I haven't added the same access for you because I haven't seen any substantial review or PR activity from you suggesting you need to do the same, but I appreciate your contribution to the push to keep the project alive just as much.

The avahi-project github organisation you created I removed you from the member list of because I synchronised the list of contributors to be the same as I added to this repository, in preparation for at some point moving the repository over but not yet. As mentioned above, there is no actual activity that is blocked by this at the moment so it is not causing any functional impediment but I agree it's the right long term goal.

I think we are heading in the right direction with 2 other contributors having been added to the project. Let's give @pemensik and @evverx all the support they need - I imagine the best kind of support there would be reviewing any open PRs and giving it a +1/-1 with details of what you checked/tested - making it easier/faster for any of us to look and merge those. Anyone is welcome to do so!

**Comment by Neustradamus** [comment]
Dear @avahi-project team members, @lathiat, @evverx, @pemensik,

Any progress for the transfer of this avahi repository (https://github.com/lathiat/avahi) and nss-mdns repository (https://github.com/lathiat/nss-mdns) to avahi-project GitHub organization that I have created to have a best future, a best development, new releases?
Note: I have not rights now.

Badly, at this time, a lot of people have problem because it is unsecure project.
- https://github.com/lathiat/avahi/issues/325
- https://github.com/lathiat/avahi/issues/430
- https://github.com/lathiat/avahi/issues/503
- https://github.com/lathiat/nss-mdns/issues/85
- https://github.com/lathiat/nss-mdns/issues/92

Some projects have been blocked at an old place because the "original" repository can not be transfered because a death, away dev, ...

For example:
- https://github.com/glennrp

Thanks in advance.

cc: @agoode.

**Comment by pemensik** [comment]
@Neustradamus Yes, actually there is a progress. @lathiat have succesufully obtained name https://github.com/avahi, so instead of @avahi-project it should move to @avahi project.

With several vulnerabilities recent vulnerabilities fixed, we have just few bugs marked important, which I would like to solve before release. Not sure we can fix name conflicts soon, because we have neither good reproducer or fix not breaking some uses. I would start a move once we release 0.9 out with all bugfixes.

**Comment by Neustradamus** [comment]
@pemensik: Nice but now it will be better to rename the current avahi organization to avahi-temp and rename avahi-project to avahi (and avahi-temp to avahi-project) and of course after transfer the 2 repositories into the good organization.

You can already create a build without all fixes.
It is easy to release several builds... better than nothing.

**Comment by evverx** [comment]
> it will be better to rename the current avahi organization to avahi-temp and rename avahi-project to avahi (and avahi-temp to avahi-project)

I wonder why anyone would do that? Is this stuff generated by ChatGPT or something like that?

> It is easy to release several builds

It isn't easy. If you think it's easy please start writing the release notes by going over https://github.com/lathiat/avahi/compare/v0.8...master. Whoever is going to cut the next release would probably be happy to accept that contribution.

**Comment by wookey** [comment]
Rereading this thread in the light of the social pressure used in the xz-utils attack is sobering. This [Neustradamus](https://github.com/Neustradamus) account also pressured for a liblzma update to 5.6.0 https://github.com/microsoft/vcpkg/issues/37197 as well as these strange requests to use their particular avahi-project and more hectoring on  ifupdown-ng  https://github.com/ifupdown-ng/ifupdown-ng/issues/234
@lathiat has probably been wise not to give them commit rights so far.
Neustradamus. If you are not a bad actor, you need to chill. Currently your pattern of behaviour across several projects looks way too much like the pattern followed by accounts used to pressure the xz-maintainer: Full explanation and links in this LWN article: https://lwn.net/SubscriberLink/967866/3c8b2ef668a62b90/
I know slow progress can be frustrating, but we all just got a bit more paranoid. Please play nice.

**Comment by Neustradamus** [comment]
@wookey: One more attack against me!

Please do not mix!

I am happy, you can see some projects which have security problems and project team who does not create release builds with CVE fixes, example here, Avahi has a lot of CVEs, you can look previous links ;)

I have already commented here after attacks against me from @gowthamgts and others:
- https://github.com/duesee/imap-flow/issues/96#issuecomment-2028014934 + https://github.com/duesee/imap-flow/issues/96#issuecomment-2028044795
- https://github.com/marlam/msmtp/issues/36#issuecomment-2028009771 + https://github.com/marlam/msmtp/issues/36#issuecomment-2028045521
- https://github.com/microsoft/vcpkg/pull/37199#issuecomment-2028009575 + https://github.com/microsoft/vcpkg/pull/37199#issuecomment-2028082689
- https://github.com/microsoft/vcpkg/pull/37841#issuecomment-2028013461 + https://github.com/microsoft/vcpkg/pull/37841#issuecomment-2028083479
- https://github.com/microsoft/vcpkg/issues/37839#issuecomment-2028085163
- https://news.ycombinator.com/item?id=39873571

You can follow my announcements here:
- https://mastodon.social/@neustradamus
- https://twitter.com/neustradamus
- https://bsky.app/profile/neustradamus.bsky.social
- https://news.ycombinator.com/user?id=neustradamus
- ...

The good point, people speak about:

The unsecure Avahi project here, with a lot of CVEs :)

SCRAM "Salted Challenge Response Authentication Mechanism" security ;)
- https://github.com/scram-sasl/info/issues/1

Badly, some people or projects like only old unsecure mechanisms, some would like security improvements.

Recall: I AM NOT LINKED TO XZ PROJECT, PLEASE LOOK ALL MY ANNOUNCEMENTS AND UPDATE REQUESTS SINCE SEVERAL YEARS.

**Comment by kaniini** [comment]
@Neustradamus I think the point is that badgering people to do work on your terms rather than our own terms carries some potential risk of maintainer error (such as giving rights to a threat actor).  I would also suggest that shitting on a project (for example: "the unsecure Avahi project here, with a lot of CVEs :)") is unlikely to make maintainers want to collaborate with you.

Maybe you are just a person who is enthusiastic about technology.  Maybe you're sitting in an alphabet soup agency office somewhere.  I certainly don't know, and don't wish to make that kind of value judgement about you, but I will say that trying to push things along right after we found at least one backdoor from a threat actor is probably not helpful to anyone.

**Comment by Neustradamus** [comment]
@kaniini: About your ifupdown-ng project, the problem was that you have not done answers, it is for this, for example, I have helped @EasyNetDev to have your attention and answers from you ;)

About Avahi project, the problem is not recent, a lot of people search a solution to remove Avahi (more used in the World than ifupdown-ng) because it was not updated, etc.
After a long time, I have found the solution which has permit the creation of a new team (I am very happy).
But at this time, it is not perfect yet, always no new version with CVE fixes which have been listed here: https://github.com/avahi/avahi/issues/503
The Avahi 0.8.0 has been released (2020-02-08): https://github.com/avahi/avahi/releases/tag/v0.8
I have done a ticket for a 0.9.0 release in 2021 before CVEs from 2023: https://github.com/avahi/avahi/issues/325

I think that you have not understand, you mix like several people: I AM NOT THE AUTHOR OF THE XZ BACKDOOR, I HAVE NO LINK WITH XZ PROJECT!

I invite you to follow me here and have announcements:
- https://mastodon.social/@neustradamus
- https://twitter.com/neustradamus
- https://bsky.app/profile/neustradamus.bsky.social
- https://news.ycombinator.com/user?id=neustradamus
- https://www.reddit.com/user/Neustradamus/
- ...

**Comment by pemensik** [comment]
Calm down @Neustradamus , people are surprised how targeted attacks were proved to be made. Especially where recently change of maintainers happened are people suspicious. Yes, I thought about avahi when reading about it first. And if you were pushing for xz update without specifying why exactly it should be updated, then it is understandable. Bad luck. It seems you are present on too many places to be imaginary person. It does not help much when you use only pseudonym though.

I doubt we know real name of xz backdoor author. That process were clearly well thought and not random feat. Very unlikely names used were of any real actors involved.

Anyway, we lack a nice summary of release notes and remaining CVE fixes. Problem is that CVE issues are faster detected than being fixed. It is a price of longer period of reduced maintaining efforts. Shape of code is worse than I anticipated. I wanted to have release fixing all important CVEs, but it should not break functionality at the same time. That does not seem a simple task, maybe releasing what we already have is safer approach. And I still struggle with my primary responsibilities elsewhere, I am sorry it is taking long already. That is not because of a lack of interest.

**Comment by Neustradamus** [comment]
@pemensik: Thanks for your answer!

I am very sorry that people come here and that I always need to specify the reasons and the reality.

I take advantage of this situation... I know that you do the best to have a perfect new version but maybe you can create a 0.9.0 first and after a 0.9.1 with latest CVE fixes. It will be always better than now...

**Comment by thesamesam** [comment]
@Neustradamus I think maybe just try to take the lesson that sometimes it comes across as pushy.

**Comment by Neustradamus** [comment]
@kaniini: I thank you for your review here: https://github.com/ifupdown-ng/ifupdown-ng/pull/191#issue-1301854464 :)
I think that @EasyNetDev will be happy to see this, a long time of patience (2022-07-12), 2 years soon.

**Comment by evverx** [comment]
> The unsecure Avahi project here

It's not insecure. It's just that these days avahi is extensively tested upstream and all sorts of issues pop up in the process. When PRs are opened/merged the test suite runs it under ASan/UBSan and Valgrind while the fuzz targets (and bespoke fuzzers like dfuzzer and radamsa) throw all sorts of stuff at it. (Downstream it goes through a packet generator that I use to find all sorts of issues in `systemd-networkd` and `systemd-resolved`) It's analyzed by static analyzers like Coverity Scan and CodeQL too. Branch/tag protection rules and stuff like that have been set up to prevent other kinds of issues.

> maybe releasing what we already have is safer approach

@pemensik agreed.

> we all just got a bit more paranoid

Let me just say that today I came across a PR on OSS-Fuzz where a dnsmasq build failure was fixed (it has been failing to compile for more than a year there). The bot there said that @pemensik is a contributor so it would seem logical to CC Petr there but I guess in light of that OSS-Fuzz issue they apparently can't do that because it isn't clear whether Petr is an upstream maintainer or not and that's kind of sad that this sort of thing is happening.

**Comment by gowthamgts** [comment]
@Neustradamus for the love of God, please stop tagging everyone involved in the last week's discussion. We have already given you answers and if you don't want to accept it, I'm not sure how to respond.

**Comment by gowthamgts** [comment]
> I think the point is that badgering people to do work on your terms rather than our own terms carries some potential risk of maintainer error (such as giving rights to a threat actor). I would also suggest that shitting on a project (for example: "the unsecure Avahi project here, with a lot of CVEs :)") is unlikely to make maintainers want to collaborate with you.

> Maybe you are just a person who is enthusiastic about technology. Maybe you're sitting in an alphabet soup agency office somewhere. I certainly don't know, and don't wish to make that kind of value judgement about you, but I will say that trying to push things along right after we found at least one backdoor from a threat actor is probably not helpful to anyone.

exactly what we also said last week [here](https://github.com/duesee/imap-flow/issues/96#issuecomment-2028024405).
